src/main/kotlin/com/google/container/tools/diagnostics/GoogleFeedbackErrorReporter.kt

```diff
@@ -0,0 +1,214 @@
+            }
+
+            val errorCallback: (Exception) -> Unit = { ex ->
+                val message =
+                    ErrorReporterBundle.message("error.googlefeedback.error", ex.message ?: "")
+                ReportMessages.GROUP
+                    .createNotification(
+                        ReportMessages.ERROR_REPORT,
+                        message,
+                        NotificationType.ERROR,
+                        NotificationListener.URL_OPENING_LISTENER
+                    )
+                    .setImportant(false)
+                    .notify(project)
+            }
+            val task = GoogleFeedbackTask(
+                project,
+                ErrorReporterBundle.message("error.googlefeedback.submitting.error.message"),
+                true,
+                event.throwable,
+                params,
+                event.message ?: "",
+                description ?: "",
+                ApplicationInfo.getInstance().fullVersion,
+                successCallback,
+                errorCallback
+            )
+            if (project == null) {
+                task.run(EmptyProgressIndicator())
+            } else {
+                ProgressManager.getInstance().run(task)
+            }
+            return true
+        }
+
+        @VisibleForTesting
+        fun buildKeyValuesMap(
+            event: IdeaLoggingEvent,
+            description: String?,
+            lastActionId: String,
+            intelliJAppNameInfo: ApplicationNamesInfo,
+            intelliJAppExtendedInfo: ApplicationInfoEx,
+            application: Application
+        ): Map<String, String> {
+
+            return mapOf(
+                // required parameters
+                ERROR_MESSAGE_KEY to nullToNone(event.message),
+                ERROR_STACKTRACE_KEY to nullToNone(getStacktrace(event)),
+                // end or required parameters
+                ERROR_DESCRIPTION_KEY to nullToNone(description),
+                LAST_ACTION_KEY to nullToNone(lastActionId),
+                OS_NAME_KEY to SystemInfo.OS_NAME,
+                JAVA_VERSION_KEY to SystemInfo.JAVA_VERSION,
+                JAVA_VM_VENDOR_KEY to SystemInfo.JAVA_VENDOR,
+                APP_NAME_KEY to intelliJAppNameInfo.fullProductName,
+                APP_CODE_KEY to intelliJAppExtendedInfo.packageCode!!,
+                APP_NAME_VERSION_KEY to intelliJAppExtendedInfo.versionName,
+                APP_EAP_KEY to java.lang.Boolean.toString(intelliJAppExtendedInfo.isEAP),
+                APP_INTERNAL_KEY to java.lang.Boolean.toString(application.isInternal),
+                APP_VERSION_MAJOR_KEY to intelliJAppExtendedInfo.majorVersion,
+                APP_VERSION_MINOR_KEY to intelliJAppExtendedInfo.minorVersion,
+                PLUGIN_VERSION to nullToNone(getPluginVersion()))
+        }
+
+        private fun getStacktrace(event: IdeaLoggingEvent): String? {
+            return if (event.throwable != null) {
```

Comment:
can we use elvis operator here for clarity?

Reply:
and probably can do this as single expression function too

Reply:
I'm not sure if I can since its something like this:

`event.throwable ?: null` 

means return `event.throwable` if its not null, and return null otherwise. We need to return `ExceptionUtil.getThrowable(..)`. Did you have something else in mind?

Reply:
Ah I see, different return type. In this case I think `?.let` comes very handy, how about:
```
        private fun getStacktrace(event: IdeaLoggingEvent): String? =
            event.throwable?.let { ExceptionUtil.getThrowableText(it) }
```?

Reply:
ah..nice. done.